File: bazel_ros2_rules/ros2/resources/rmw_isolation/test/isolated_talker.cc
#include <chrono>
#include <memory>

#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>

#include "rmw_isolation/rmw_isolation.h"

class IsolatedTalker : public rclcpp::Node {
 public:
  IsolatedTalker(const std::string& uuid)
      : rclcpp::Node("isolated_talker"), uuid_(uuid) {
    publisher_ = this->create_publisher<std_msgs::msg::String>("uuid", 10);
    timer_ = this->create_wall_timer(
        std::chrono::duration<double>(0.1),
        std::bind(&IsolatedTalker::timer_callback, this));
  }

 private:
  void timer_callback() const {
    std_msgs::msg::String message;
    message.data = uuid_;
    publisher_->publish(message);
  }

  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
  std::string uuid_;
};

int main(int argc, char* argv[]) {
  const char* TEST_TMPDIR = std::getenv("TEST_TMPDIR");
  if (TEST_TMPDIR != nullptr) {
    ros2::isolate_rmw_by_path(argv[0], TEST_TMPDIR);
  }
  rclcpp::init(argc, argv);
  std::string uuid{TEST_TMPDIR != nullptr ? TEST_TMPDIR : "none"};
  rclcpp::spin(std::make_shared<IsolatedTalker>(uuid));
  rclcpp::shutdown();
  return 0;
}

File: bazel_ros2_rules/ros2/resources/rmw_isolation/test/isolated_talker.py
import os

import rclpy
import rclpy.node
import std_msgs.msg


class IsolatedTalker(rclpy.node.Node):

    def __init__(self, uuid):
        super().__init__('isolated_talker')
        self._publisher = self.create_publisher(
            std_msgs.msg.String, 'uuid', 10)
        self._timer = self.create_timer(0.1, self._timer_callback)
        self._uuid = uuid

    def _timer_callback(self):
        msg = std_msgs.msg.String()
        msg.data = self._uuid
        self._publisher.publish(msg)


def main():
    if 'TEST_TMPDIR' in os.environ:
        from rmw_isolation import isolate_rmw_by_path
        isolate_rmw_by_path(os.environ['TEST_TMPDIR'])

    rclpy.init()

    uuid = os.environ.get('TEST_TMPDIR', 'none')
    try:
        rclpy.spin(IsolatedTalker(uuid))
    # TODO(hidmic): simplify `except` blocks as follows:
    #
    #   except KeyboardInterrupt:
    #       pass
    #   finally:
    #       rclpy.try_shutdown()
    #
    # when rclpy 3.2.0 is rolled out (or any rclpy version
    # including https://github.com/ros2/rclpy/pull/868 is).
    except rclpy.executors.ExternalShutdownException:
        pass
    except KeyboardInterrupt:
        rclpy.shutdown()
    except Exception:
        rclpy.shutdown()
        raise


if __name__ == '__main__':
    main()
